File: src/emulator/GameBoy.cs
using System;

namespace DMSharp
{

    public class GameBoy
    {


        public CPU cpu;
        public GPU gpu;
        public InterruptController interrupts;
        public MemoryBus bus;
        public SoundChip soundChip;

        public Timer timer;

        public int speedMul = 1;

        public GameBoy()
        {
            this.cpu = new CPU(this);
            this.gpu = new GPU(this);
            this.bus = new MemoryBus(this);
            this.soundChip = new SoundChip(this);
            this.timer = new Timer(this);
            this.interrupts = new InterruptController(this.bus);
            Util.WriteDebug("New gameboy!");
            // setInterval(() => { this.bus.ext.saveGameSram(); }, 1000);
        }


        internal void Step()
        {
            this.soundChip.Step();
            this.cpu.Step();
            this.gpu.Step();
            this.timer.Step();
        }


        internal void speedStop()
        {
            this.cpu.stopNow = true;
            // this.soundChip.tjs.setMuted(true);
        }

        internal void speed()
        {
            this.cpu.debugging = false;
            this.cpu.stopNow = false;
            // this.soundChip.tjs.setMuted(false);
        }

        internal void frame()
        {
            long i = 0;
            // const max = 70224; // Full frame GPU timing
            var max = 70224 * this.speedMul; // Full frame GPU timing, double speed
            if (this.cpu.breakpoints.Contains(this.cpu.pc) || this.cpu.stopNow)
            {
                this.speedStop();
            }
            while (i < max && !this.cpu.breakpoints.Contains(this.cpu.pc) && !this.cpu.stopNow)
            {
                this.Step();
                i += this.cpu.lastInstructionCycles;
            }
            if (this.cpu.stopNow) this.cpu.stopNow = false;
        }



        public void Reset()
        {
            this.cpu.Reset();
            this.gpu.Reset();
            this.interrupts.Reset();
            this.bus.Reset();
            this.timer.Reset();
            this.soundChip.Reset();
        }
    }
}

File: src/gameboy/core/gameboy.ts
import CPU from './cpu/cpu';
import GPU from './video/gpu';
import MemoryBus from './memory/memorybus';
import Disassembler from '../tools/disassembler';
import { writeDebug } from '../tools/debug';
import SoundChip from './sound/sound';
import Timer from './components/timer';
import Decoder from './cpu/decoder';

export default class GameBoy {
    cpu = new CPU(this);
    gpu = new GPU(this);
    bus = new MemoryBus(this);

    cgb = false;
    doubleSpeed = false;
    prepareSpeedSwitch = false;

    soundChip = new SoundChip(this);

    timer = new Timer(this);

    constructor(cgb: boolean) {
        writeDebug("New gameboy!");
        this.cgb = cgb;
        setInterval(() => { this.bus.ext.saveGameSram(); }, 1000);
    }

    step() {
        this.cpu.step();
        this.timer.step();

        if (this.doubleSpeed) {
            this.cpu.step();
            this.timer.step();
        }

        this.soundChip.step();
        this.gpu.step();
    }

    speedMul = 1;
    speedIntervals: Array<number> = [];

    speedStop() {
        this.speedIntervals.forEach(i => { clearInterval(i); });
        this.soundChip.tjs.setMuted(true);
    }

    speed() {
        this.cpu.debugging = false;
        this.speedIntervals.push(setInterval(() => { this.frame(); }, 16));
        this.soundChip.tjs.setMuted(false);
    }

    frame() {
        let i = 0;
        // const max = 70224; // Full frame GPU timing
        const max = 70224 * this.speedMul; // Full frame GPU timing, double speed

        while (i < max) {
            if (this.cpu.breakpoints[this.cpu.pc]) {
                this.speedStop();
                return;
            }
            this.step();
            i += this.cpu.lastInstructionCycles;
        }
    }



    reset() {
        this.cpu.reset();
        this.gpu.reset();
        this.bus.interrupts.reset();
        this.bus.reset();
        this.timer.reset();
        this.soundChip.reset();

        this.doubleSpeed = false;
        this.prepareSpeedSwitch = false;
    }
}

export { GameBoy, CPU, GPU, MemoryBus, Timer, Disassembler, Decoder }; 